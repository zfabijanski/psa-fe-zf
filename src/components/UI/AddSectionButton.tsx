import React from "react";
import styled from "styled-components/macro";
import { useIntl } from "react-intl";
import { Icon, TIconType } from "./Icon";
import { TypoH4 } from "./typography";

export interface IAddSectionButtonProps {
  className?: string;
  textTrKey: string;
  buttonText?: string;
  buttonIconName?: TIconType;
  onClick: () => void;
}

export const AddSectionButton: React.FC<IAddSectionButtonProps> = (props) => {
  const intl = useIntl();

  return (
    <Wrapper className={props.className} onClick={props.onClick}>
      <StyledTypoH4>{intl.formatMessage({ id: props.textTrKey })}</StyledTypoH4>
      <Separator />
      <CustomButton>
        {intl.formatMessage({ id: props.buttonText! })}
        <CustomButtonIcon width={20} name={props.buttonIconName!} />
      </CustomButton>
    </Wrapper>
  );
};

AddSectionButton.defaultProps = {
  buttonText: "common.add",
  buttonIconName: "plus",
};

const StyledTypoH4 = styled(TypoH4)`
  color: ${({ theme }) => theme.newColors.primary100};
`;

const Separator = styled.div`
  height: 1px;
  background-color: ${({ theme }) => theme.newColors.primary100};
  flex: 1;
  min-width: 200px;
  margin: 0 12px;
`;

const CustomButton = styled.div`
  background-color: ${({ theme }) => theme.newColors.white100};
  border: 2px solid ${({ theme }) => theme.newColors.primary100};
  color: ${({ theme }) => theme.newColors.primary100};
  padding: 0 24px;
  height: 44px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  font-weight: 700;
`;

const CustomButtonIcon = styled(Icon)`
  margin-left: 12px;

  svg g {
    &,
    path,
    circle,
    rect {
      stroke: ${({ theme }) => theme.newColors.primary100} !important;
    }

    rect {
      fill: ${({ theme }) => theme.newColors.primary100} !important;
    }

    tspan {
      stroke: transparent !important;
      fill: ${({ theme }) => theme.newColors.primary100} !important;
    }
  }

  svg > path {
    stroke: ${({ theme }) => theme.newColors.primary100} !important;
  }
`;

const Wrapper = styled.div<{ justify?: string }>`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  position: relative;

  &:hover:not(:active) {
    ${StyledTypoH4} {
      color: ${({ theme }) => theme.newColors.gray100};
    }

    ${Separator} {
      background-color: ${({ theme }) => theme.newColors.gray100};
    }

    ${CustomButton} {
      border-color: ${({ theme }) => theme.newColors.gray100};
      color: ${({ theme }) => theme.newColors.gray100};
    }

    ${CustomButtonIcon} {
      svg g {
        &,
        path,
        circle,
        rect {
          stroke: ${({ theme }) => theme.newColors.gray100} !important;
        }

        rect {
          fill: ${({ theme }) => theme.newColors.gray100} !important;
        }

        tspan {
          stroke: transparent !important;
          fill: ${({ theme }) => theme.newColors.gray100} !important;
        }
      }

      svg > path {
        stroke: ${({ theme }) => theme.newColors.gray100} !important;
      }
    }
  }

  &:active {
    ${CustomButton} {
      background-color: ${({ theme }) => theme.newColors.primary10};
    }
  }
`;
