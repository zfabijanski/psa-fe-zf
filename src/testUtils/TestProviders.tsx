import { render, RenderOptions } from "@testing-library/react";
import { FC, PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { Normalize } from "styled-normalize";
import GlobalStyles from "theme/globalStyles";
import { theme } from "theme/theme";
import rootStore, { configureStore, rootReducer as reducer } from "AppStore";
import { BrowserRouter } from "react-router-dom";
import { Modal } from "components/UI/Modal";
import { IntlProvider } from "react-intl";
import { plMessages } from "translations/pl";
import I18nProvider from "I18nProvider";

interface TestProvidersProps {
  store?: typeof rootStore;
  renderModal?: boolean;
  realIntlProvider?: boolean;
}

export const addTestProviders =
  ({
    store = configureStore({ reducer }),
    renderModal = true,
    // @TODO: update tests to use realIntlProvider everywhere
    realIntlProvider = false,
  }: TestProvidersProps): FC<PropsWithChildren<{}>> =>
  ({ children }) => {
    const TranslationsProvider = realIntlProvider ? IntlProvider : I18nProvider;

    return (
      <Provider store={store}>
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <TranslationsProvider locale="pl" messages={plMessages}>
              <Normalize />
              <GlobalStyles />
              {children}
              {renderModal && <Modal />}
            </TranslationsProvider>
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    );
  };

const customRender = (
  ui: ReactElement,
  options: Omit<RenderOptions, "options"> = {},
  testProviderOptions: TestProvidersProps = {}
) => render(ui, { wrapper: addTestProviders(testProviderOptions), ...options });

export * from "@testing-library/react";

// override render method
export { customRender as render };
