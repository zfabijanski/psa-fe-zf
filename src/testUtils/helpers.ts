export const sleep = (ms: number = 100): Promise<unknown> =>
  new Promise((resolve) => setTimeout(resolve, ms));
