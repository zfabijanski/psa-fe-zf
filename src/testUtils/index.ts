export * from "./TestProviders";
export * from "./mockCreators";
export * from "./helpers";
