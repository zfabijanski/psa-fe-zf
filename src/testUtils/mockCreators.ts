import {
  ICalculatorCoversConfig,
  IProductCovers,
} from "features/Calculator/types";
import {
  DocumentState,
  IDirectory,
  IDirectoryMetadata,
  IDocumentMetadata,
} from "features/Library/types";
import { IMeeting } from "features/Meeting/History/types";
import { IIllustration, ProductGuid } from "features/Products/types";
import { IModalData, ModalTypes } from "models/Modal";
import {
  Frequency,
  IFundRiskProfileConfig,
  IProductConfig,
  IProductFund,
  IProductGuarantee,
  InvestmentProfile,
  Role,
} from "models/common";
import {
  CalculationStatus,
  Direction,
  ICalculationDetailsDto,
  ICover,
  IInsured,
  YesNo,
} from "models/calculator";

export const createCalculationDetailsDto = (
  overrides?: Partial<ICalculationDetailsDto>
): ICalculationDetailsDto => ({
  calculation_id: 1,
  working_date: "2020-01-01",
  premiums_total: 100,
  guaranteed_death_benefit: 100,
  guaranteed_maturity_benefit: 100,
  projected_maturity_value_egr_m: 100,
  product_categories: [],
  uuid: "uuid",
  version: "version",
  source_system: "source_system",
  product_guid: 1,
  insureds_list: [],
  covers_list: [],
  funds_list: [],
  ...overrides,
});

export const createInsured = (overrides?: Partial<IInsured>): IInsured => ({
  insured_idx: 1,
  insured_uuid: "insured_uuid",
  age_in_months: 1,
  date_of_birth: "2020-01-01",
  role: Role.MAIN,
  ...overrides,
});

export const createCover = (overrides?: Partial<ICover>): ICover => ({
  cover_id: 1,
  cover_idx: 1,
  insured_idx: 1,
  code: "code",
  frequency_code: Frequency.Annually,
  payments: 1,
  direction_of_calculation: Direction.SumAssuredToPremium,
  investment_profile: InvestmentProfile.Balanced,
  indexation: YesNo.Yes,
  sum_assured: 1,
  premium_per_freq: 1,
  duration: 1,
  idx_option: YesNo.Yes,
  rop_option: YesNo.Yes,
  add_sum_assured: 1,
  ending_age: 1,
  exp_mat_ben: 1,
  ...overrides,
});

export const createCalculatorCoversConfig = (
  overrides?: Partial<ICalculatorCoversConfig>
): ICalculatorCoversConfig => ({
  coverName: "coverName",
  isChild: false,
  isHospital: false,
  ...overrides,
});

export const createCalculatorCoversConfigMap = (
  overrides?: Partial<Record<string, ICalculatorCoversConfig>>
): Record<string, ICalculatorCoversConfig> => ({
  coverName1: createCalculatorCoversConfig({ coverName: "coverName1" }),
  coverName2: createCalculatorCoversConfig({ coverName: "coverName2" }),
  ...overrides,
});

export const createProductCovers = (
  overrides?: Partial<IProductCovers>
): IProductCovers => ({
  mainCover: "mainCover",
  additionalCovers: ["additionalCover1", "additionalCover2"],
  wopCovers: ["wopCover1", "wopCover2"],
  childCovers: ["childCover1", "childCover2"],
  ...overrides,
});

export const createProductCoversMap = (
  overrides?: Partial<Record<string, IProductCovers>>
): Record<string, IProductCovers> => ({
  mainCover1: createProductCovers({ mainCover: "mainCover1" }),
  mainCover2: createProductCovers({ mainCover: "mainCover2" }),
  ...overrides,
});

export const createProductConfig = (
  overrides?: Partial<IProductConfig>
): IProductConfig => ({
  product_guid: 1,
  product_name: "product_name",
  product_categories: [],
  product_freqs: [],
  product_covers: [],
  product_funds: [],
  ...overrides,
});

export const createProductConfigMap = (
  overrides?: Partial<Record<string, IProductConfig>>
): Record<string, IProductConfig> => ({
  1: createProductConfig({ product_name: "product_name1" }),
  2: createProductConfig({ product_name: "product_name2" }),
  ...overrides,
});

export const createFundRiskProfileConfig = (
  overrides?: Partial<IFundRiskProfileConfig>
): IFundRiskProfileConfig => ({
  code: "code",
  name_pl: "name_pl",
  name_en: "name_en",
  risk_level: 1,
  status_code: "status_code",
  notes: "notes",
  ...overrides,
});

export const createProductGuarantee = (
  overrides?: Partial<IProductGuarantee>
): IProductGuarantee => ({
  guarantee_code: "guarantee_code",
  guarantee_name: "guarantee_name",
  guarantee_min_value: "1",
  guarantee_duration_months: "4",
  guarantee_percentages: [{ guarantee_percent: 10 }],
  ...overrides,
});

export const createProductFund = (
  overrides?: Partial<IProductFund>
): IProductFund => ({
  cover_code: "cover_code",
  fund_code: "fund_code",
  fund_name: "fund_name",
  fund_risk_profile: "code",
  fund_order: 1,
  guarantees: [createProductGuarantee()],
  ...overrides,
});

export const createModalData = (
  overrides: Partial<IModalData> = {}
): IModalData => ({
  modalIllustration: "alert-fill",
  modalTitleTrKey: "modalTitleTrKey",
  modalContentTrKey: "modalContentTrKey",
  modalContent: "modalContent",
  modalButtons: {
    confirm: { textTrKey: "modal.ok" },
  },
  modalType: ModalTypes.default,
  maskClosable: true,
  ...overrides,
});

export const createIllustration = (
  overrides: Partial<IIllustration> = {}
): IIllustration => ({
  idd_id: 1,
  product_guid: ProductGuid.Savings,
  adequacy_id: 1,
  order_id: 1,
  expected_amount: { value: 3 },
  premium_total: { value: 3 },
  duration: { value: 3 },
  premium_main_cover: { value: 3 },
  frequency: { value: 3 },
  sum_of_premiums: { value: 3 },
  sum_assured: { value: 3 },
  investment_profile: { value: 3 },
  pru_calc_calculation_id: 1,
  status: CalculationStatus.Illustrated,
  json_needs_analysis: null,
  ...overrides,
});

export const createDirectory = (
  overrides?: Partial<IDirectory>
): IDirectory => ({
  directoryId: 1,
  directories: [],
  documentsIds: [1, 2, 3],
  ...overrides,
});

export const createDirectoryMetadata = (
  overrides?: Partial<IDirectoryMetadata>
): IDirectoryMetadata => ({
  directoryId: 1,
  name: "name",
  isOpen: false,
  ...overrides,
});

export const createDocumentMetadata = (
  overrides?: Partial<IDocumentMetadata>
): IDocumentMetadata => ({
  documentId: 1,
  guid: "guid",
  name: "name",
  state: DocumentState.NotAttached,
  isStateChanging: false,
  moduleName: "PRU_LIBRARY",
  ...overrides,
});

export const createMeeting = (overrides?: Partial<IMeeting>): IMeeting => ({
  meetingNo: "meetingNo",
  meetingId: 1,
  name: "name",
  age: 18,
  lastAccessDate: new Date("2023-01-01"),
  createDate: new Date("2023-01-01"),
  statusId: 1,
  statusName: "statusName",
  statusDescription: "statusDescription",
  state: "state",
  haveAdequacy: false,
  haveCalculation: false,
  haveIdd: false,
});
