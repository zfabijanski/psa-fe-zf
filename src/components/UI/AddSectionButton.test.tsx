import { AddSectionButton, IAddSectionButtonProps } from "./AddSectionButton";
import { render, screen } from "testUtils";

const mockOnClick = jest.fn();
const defaultProps: IAddSectionButtonProps = {
  textTrKey: "test",
  onClick: mockOnClick,
};

const setup = (props: Partial<IAddSectionButtonProps> = {}) =>
  render(<AddSectionButton {...defaultProps} {...props} />);

describe("AddSectionButton", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should render the textTrKey", () => {
    setup();
    expect(screen.getByText("test")).toBeInTheDocument();
  });

  it("should render the buttonText when provided", () => {
    setup({ buttonText: "buttonText" });
    expect(screen.getByText("buttonText")).toBeInTheDocument();
  });

  it("should render the buttonIconName when provided", () => {
    setup({ buttonIconName: "alert-double-triangle" });
    expect(screen.getByText("alert-double-triangle.svg")).toBeInTheDocument();
  });

  it("should call onClick when clicked", () => {
    setup();
    screen.getByText("test").click();
    expect(mockOnClick).toHaveBeenCalled();
  });
});
