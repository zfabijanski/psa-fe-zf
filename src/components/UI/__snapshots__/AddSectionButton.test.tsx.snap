// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSectionButton should render correctly 1`] = `
.c8 {
  display: inline-block;
  cursor: inherit;
}

.c8 > svg {
  width: 20px;
  height: 100%;
  display: block;
}

.c2 {
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
}

.c3 {
  color: #055A60;
}

.c5 {
  height: 1px;
  background-color: #055A60;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 200px;
  margin: 0 12px;
}

.c7 {
  background-color: #FFFFFF;
  border: 2px solid #055A60;
  color: #055A60;
  padding: 0 24px;
  height: 44px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 700;
}

.c10 {
  margin-left: 12px;
}

.c10 svg g,
.c10 svg g path,
.c10 svg g circle,
.c10 svg g rect {
  stroke: #055A60 !important;
}

.c10 svg g rect {
  fill: #055A60 !important;
}

.c10 svg g tspan {
  stroke: transparent !important;
  fill: #055A60 !important;
}

.c10 svg > path {
  stroke: #055A60 !important;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
}

.c0:hover:not(:active) .c1 {
  color: #2F2F2F;
}

.c0:hover:not(:active) .c4 {
  background-color: #2F2F2F;
}

.c0:hover:not(:active) .c6 {
  border-color: #2F2F2F;
  color: #2F2F2F;
}

.c0:hover:not(:active) .c9 svg g,
.c0:hover:not(:active) .c9 svg g path,
.c0:hover:not(:active) .c9 svg g circle,
.c0:hover:not(:active) .c9 svg g rect {
  stroke: #2F2F2F !important;
}

.c0:hover:not(:active) .c9 svg g rect {
  fill: #2F2F2F !important;
}

.c0:hover:not(:active) .c9 svg g tspan {
  stroke: transparent !important;
  fill: #2F2F2F !important;
}

.c0:hover:not(:active) .c9 svg > path {
  stroke: #2F2F2F !important;
}

.c0:active .c6 {
  background-color: #EDF9F9;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2 c3"
    >
      test
    </div>
    <div
      class="c4 c5"
    />
    <div
      class="c6 c7"
    >
      common.add
      <div
        class="c8 c9 c10"
      >
        <svg>
          plus.svg
        </svg>
      </div>
    </div>
  </div>
  <div />
</div>
`;
